### PR TITLE
infra: Set -ftrivial-auto-var-init=pattern for address sanitizer

### PR DESCRIPTION
**Pull request by MarcoFalke**
Reading uninitialized memory is undefined behaviour, thus non-deterministic. Non-determinism makes it harder to reproduce a bug. To fix some cases of non-determinism, I suggest to add `-ftrivial-auto-var-init=pattern` to the address sanitizer.

Obviously not all uninitialized reads can be caught by this and Valgrind or MSAN are still recommended tools to fight this class of bug.

Even though `-ftrivial-auto-var-init=pattern` will *hide* some bugs from Valgrind or MSAN, those tools are largely incompatible with ASAN, so adding the option to the ASAN flags should be fine.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
I tested with an example bug that the crash is now deterministic with a single iterations. Previously a few iterations (~8) out of the 100 max (see commit 66b19116b) were needed.

**Comment by MarcoFalke** [comment]
Happy to test more if there are any ideas. Also, I am wondering if I am missing any downside of this?

**Comment by jonathanmetzman** [comment]
Thanks @MarcoFalke for doing this. I think there will be a lot of back and forth on this.

@morehouse @kcc @inferno-chromium @oliverchang @asraa what do you think of this?

**Comment by morehouse** [comment]
I think this change is net positive.  Would like to hear other opinions though.

One negative:

- Could cause some bugs to be missed in ASan:
```c++
bool x;  // auto-inited to true
if (x) {
  ...
} else {
  // Some bug here
  ...
}
```

**Comment by kcc** [comment]
Yea, may miss some bugs. 
Also, not sure if this is going to interfere with `-fsanitize=bool`
May still be net-positive as it will allow asan to focus on bugs it can properly report.

**Comment by MarcoFalke** [comment]
It looks like `-fsanitize=bool` will actually fix the issue found by @morehouse . Currently `-fsanitize=bool` is only set for the UBSAN in oss-fuzz, but I can enable it for ASAN as well.

For testing:

```
( echo 'int main() { bool b; return int{b}; }' | clang++ -x c++ - -fsanitize=address,bool,enum -ftrivial-auto-var-init=pattern -o exe ) && ./exe 
<stdin>:1:33: runtime error: load of value 170, which is not a valid value for type 'bool'
```

(without `-ftrivial-auto-var-init=pattern`, it passes for me locally)

File: infra/base-images/base-builder/Dockerfile [hunk 1]
```diff
@@ -127,7 +127,7 @@ RUN cd $SRC/ && \
     rm -rf $SRC/jazzer
 
 # Default build flags for various sanitizers.
-ENV SANITIZER_FLAGS_address "-fsanitize=address -fsanitize-address-use-after-scope"
+ENV SANITIZER_FLAGS_address "-fsanitize=address -fsanitize-address-use-after-scope -ftrivial-auto-var-init=pattern"
```
Comment (MarcoFalke) [inline infra/base-images/base-builder/Dockerfile:130]:
```suggestion
ENV SANITIZER_FLAGS_address "-fsanitize=address,bool,enum -fsanitize-address-use-after-scope -ftrivial-auto-var-init=pattern"
```

**Comment by morehouse** [comment]
LGTM

**Comment by MarcoFalke** [comment]
Pushed a fix for `bad_build_check`. CI is green and a few local builds were also green.

**Comment by MarcoFalke** [comment]
Anything left to do here? :)

**Pull request closed without merging**